pkg/build/controller/buildpod/controller.go

```diff
@@ -0,0 +1,404 @@
+		glog.V(4).Infof("Build %s/%s is a pipeline build, ignoring", build.Namespace, build.Name)
+		return nil
+	}
+	// If build was cancelled, we'll leave HandleBuild to update the build
+	if build.Status.Cancelled {
+		glog.V(4).Infof("Cancelation for build %s/%s was already triggered, ignoring", build.Namespace, build.Name)
+		return nil
+	}
+
+	if buildutil.IsBuildComplete(build) {
+		glog.V(4).Infof("Pod was deleted but build %s/%s is already completed, so no need to update it.", build.Namespace, build.Name)
+		return nil
+	}
+
+	nextStatus := buildapi.BuildPhaseError
+	if build.Status.Phase != nextStatus {
+		glog.V(4).Infof("Updating build %s/%s status %s -> %s", build.Namespace, build.Name, build.Status.Phase, nextStatus)
+		build.Status.Phase = nextStatus
+		build.Status.Reason = buildapi.StatusReasonBuildPodDeleted
+		build.Status.Message = buildapi.StatusMessageBuildPodDeleted
+		common.SetBuildCompletionTimeAndDuration(build)
+		if err := bc.buildUpdater.Update(build.Namespace, build); err != nil {
+			return fmt.Errorf("Failed to update build %s/%s: %v", build.Namespace, build.Name, err)
+		}
+	}
+	return nil
+}
+
+func (bc *BuildPodController) worker() {
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildPodController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	pod, err := bc.getPodByKey(key.(string))
+	if err != nil {
+		utilruntime.HandleError(err)
+	}
+
+	if pod == nil {
+		return false
+	}
+
+	err = bc.HandlePod(pod)
+	bc.handleErr(err, key, pod)
+
+	return false
+}
+
+func (bc *BuildPodController) getPodByKey(key string) (*kapi.Pod, error) {
+	obj, exists, err := bc.podStore.Indexer.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve pod %q from store: %v", key, err)
+		bc.queue.Add(key)
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("Pod %q has been deleted", key)
+		return nil, nil
+	}
+
+	return obj.(*kapi.Pod), nil
+}
+
+func (bc *BuildPodController) updatePod(old, cur interface{}) {
+	// A periodic relist will send update events for all known pods.
+	curPod := cur.(*kapi.Pod)
+	oldPod := old.(*kapi.Pod)
+	if curPod.ResourceVersion == oldPod.ResourceVersion {
+		return
+	}
+	if isBuildPod(curPod) {
+		bc.enqueuePod(curPod)
+	}
+}
+
+func (bc *BuildPodController) deletePod(obj interface{}) {
+	pod, ok := obj.(*kapi.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone: %+v", obj))
+			return
+		}
+		pod, ok = tombstone.Obj.(*kapi.Pod)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a pod: %+v", obj))
+			return
+		}
+	}
+	if isBuildPod(pod) {
+		err := bc.HandleBuildPodDeletion(pod)
+		utilruntime.HandleError(err)
+	}
+}
+
+func (bc *BuildPodController) addBuild(obj interface{}) {
+	build := obj.(*buildapi.Build)
+	bc.checkBuildPodDeletion(build)
```

Comment:
When a build is created, its pod doesn't exist but we'll end up triggering a deletion event for the pod, right?